tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,1173 @@
+        item, deps = yield crawl_item_and_deps(ProductFromParent)
+        assert any([True for w in caught_warnings if msg in str(w.message)])
+
+    assert item == ProductFromParent(name="subclass product name")
+    assert_deps(deps, {"item": ProductFromParent})
+
+    # calling the actual page objects should still work
+
+    item, deps = yield crawl_item_and_deps(ParentProductPage)
+    assert item == ProductFromParent(name="parent product name")
+    assert_deps(deps, {"page": ParentProductPage})
+
+    item, deps = yield crawl_item_and_deps(SubclassProductPage)
+    assert item == ProductFromParent(name="subclass product name")
+    assert_deps(deps, {"page": SubclassProductPage})
+
+
+@attrs.define
+class PriorityProductFromParent:
+    name: str
+
+
+@handle_urls(URL, priority=600)
+class PriorityParentProductPage(ItemPage[PriorityProductFromParent]):
+    @field
+    def name(self) -> str:
+        return "priority parent product name"
+
+
+@handle_urls(URL)
+class PrioritySubclassProductPage(PriorityParentProductPage):
+    @field
+    def name(self) -> str:
+        return "priority subclass product name"
+
+
+@inlineCallbacks
+def test_item_return_subclass_priority() -> None:
+    """Same case as with ``test_item_return_subclass()`` but now the parent PO
+    uses a higher priority of 600 than the default 500.
+    """
+    item, deps = yield crawl_item_and_deps(PriorityProductFromParent)
+    assert item == PriorityProductFromParent(name="priority parent product name")
+    assert_deps(deps, {"item": PriorityProductFromParent})
+
+    # calling the actual page objects should still work
+
+    item, deps = yield crawl_item_and_deps(PriorityParentProductPage)
+    assert item == PriorityProductFromParent(name="priority parent product name")
+    assert_deps(deps, {"page": PriorityParentProductPage})
+
+    item, deps = yield crawl_item_and_deps(PrioritySubclassProductPage)
+    assert item == PriorityProductFromParent(name="priority subclass product name")
+    assert_deps(deps, {"page": PriorityParentProductPage})
+
+
+@attrs.define
+class ReplacedProduct:
+    name: str
+
+
+@handle_urls(URL, to_return=ReplacedProduct)
+class ReplacedProductPage(ItemPage[Product]):
+    @field
+    def name(self) -> str:
+        return "replaced product name"
+
+
+@inlineCallbacks
+def test_item_to_return_in_handle_urls(caplog) -> None:
+    """Even if ``@handle_urls`` could derive the value for the ``to_return``
+    parameter when the class inherits from something like ``ItemPage[ItemType]``,
+    any value passed through its ``to_return`` parameter should take precedence.
+
+    Note that that this produces some inconsistencies between the rule's item
+    class vs the class that is actually returned. Using the ``to_return``
+    parameter in ``@handle_urls`` isn't recommended because of this.
+
+    This also causes an ``UndeclaredProvidedTypeError`` since the item provider
+    has received a different type of item class from the page object.
```

Comment:
I think the expected behavior for to_return should be to "lie", and actually return ReplacedProduct, not fail. I'm fine with not supporting it, but then probably we should have an xfail test, not a test which enforces the current UndeclaredProvidedTypeError.

Reply:
Updated 3 variations of this test with an xfail. https://github.com/scrapinghub/scrapy-poet/pull/88/commits/4d25d8c7c59c9f075238acfcf1fbdb5f37b9c91a